### ISSUE TITLE
[rush] Deploy by crawling the dependencies of a project and copying this subset of files

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#1078]
@SoyYoRafa [recently asked](https://gitter.im/web-build-tools/web-build-tools?at=5c46a7000a491251e335a78a) for a way to extract the subset of installed dependencies for an application project, in order to deploy by copying these files:

> Hi, I am porting a project to use Rush. I am curious what is the preferred way to deploy/release an application to production? The main concern is reproducing in production the dependencies used in development. If we deploy via npm/private artifactory, there isn't a package-lock.json so we can't guarantee that. We have been deploying the applications with tarballs but this doesn't work per-application. with Rush/pnpm. We would have to tarball the entire monorepo. Also, our production environment is air gapped and doesn't have access to a npm registry. Is there a way to not "hoist"?

The discussion ended up with [a proposal](https://gitter.im/web-build-tools/web-build-tools?at=5c49e561cb47ec3000964bf2) for a dedicated Rush command:

> Here's another idea: what if we created a new command (e.g. `rush copy-deps`?) that would crawl the node_modules folder for your application and copy its subset of dependencies into an isolated target folder including symlinks. Then you could copy that folder to your web server using `tar` or `rsync` which can preserve symlinks
> To clarify the "symlinks" here would not be pointing into the monorepo **common/temp/node_modules** folder, but rather a local folder that is part of the isolated subset, and which e.g. does not have the NPM tarball cache.

This seems rather easy to implement. It would complement the other ideas from https://github.com/Microsoft/web-build-tools/issues/904.

**Comment by octogonz** [comment]
This same "crawling all dependencies" algorithm might also help with @sstchurms's question about [reporting the dependency tree for a project](https://gitter.im/web-build-tools/web-build-tools?at=5c5a1ef9454aad4df7ff8e12):

> Hi there, wondering if there is a way to input a package name and get a list of all packages in my project that depend on it?  With traditional npm, I could use:  npm ls packageName, but this seems not to work with Rush
> Any suggestions as to how I can do something similar?  Or something similar to npm audit?
> My project is being flagged for a boat load of vulnerable packages, but I don't depend on any of them directly, so determining what I really do depend on so I can update it, is proving to be infuriating

**Comment by octogonz** [comment]
My basic idea was something like this (pseudocode):

```js
import resolve from 'resolve';  // https://www.npmjs.com/package/resolve

function collectFolders(folderPathSet: Set<string>, packageJsonPath: string): void {
  if (!folderPathSet.has(packageJsonPath)) {
    folderPathSet.add(packageJsonPath);
    
    const packageJson = loadPackageJson(packageJsonPath);
    
    foreach (dependencyPackageName in packageJson.dependencies) {

      const resolvedDependency = resolve.sync(packageName,
        { basedir: path.dirname(packageJsonPath) });

      const dependencyPackageJsonPath: string 
        = fs.realpathSync(resolvedDependency.packageJsonPath);

      collectFolders(folderPathSet, dependencyPackageJsonPath)
    }
    
  }   
}

const folderPathSet = new Set<string>();

// Start with the package.json in the current folder, and then resolve all its dependencies
collectFolders(folderPathSet, "./package.json");

// Now deploy all files whose path is under one of the entries from folderPathSet
```

**Comment by SoyYoRafa** [comment]
Any thoughts on `rush copy` which copies the package without node_modules and then copies the subset of node_modules? Otherwise, I would imagine most people using `rush copy-deps` would have to do this themselves.

**Comment by octogonz** [comment]
> We would only be interested in copying production dependencies.

How would the algorithm identify production dependencies?

From **package.json** the only information we have is have:
- dependencies
- devDependencies
- peerDependencies
- optionalDependencies

**Comment by octogonz** [comment]
> "dependencies" are the production dependencies in npm (e.g. npm install --save-prod). The main idea is to avoid copying dev dependencies such as tsc, tslint, mocha, chai, etc that are not needed for deployment.

According to the [manual](https://docs.npmjs.com/cli/install), `--save-prod` simply means to store it in `dependencies` (instead of e.g. `devDependencies`).  So the above algorithm already implements that idea.

@SoyYoRafa maybe you could try running this script on one of your projects and see if the results look correct?

@sstchurms you could also try it and see how close the  output is to accomplishing the `npm audit` reporting you asked about.

If people think this algorithm is likely to work, then the next step would be to implement the actual copying algorithm (which involves reproducing symlinks correctly).  After validating that the result actually works correctly, then we would discuss what the user interface should be.

I'm busy with a lot of other stuff right now, so if others want to push this idea along, feel free.  :-)

**Comment by SoyYoRafa** [comment]
The output looks correct.

**Comment by poelstra** [comment]
As a 👍 from our project, we have two use-cases for something like this:
1. Generating an overview of all packages (including dev), and a similar overview of just the production packages, to be able to crawl them for OSS license information to produce a ThirdPartyNotices file, and to verify the packages against a license whitelist.
2. Generating a 'normal' `node_modules` structure such that tools like `nexe` or `pkg` understand them and can properly generate a standalone binary. Apparently, the dependency walkers in them get confused by Rush's symlink structure (or pnpm's, not sure).

One thing that seems missing from the proposed code is how to handle `peerDependencies` and `optionalDependencies`. I expect to do some experiments here to see how that could be handled in a `rush`/`pnpm` style modules folder.

**Comment by octogonz** [comment]
> One thing that seems missing from the proposed code is how to handle `peerDependencies` and `optionalDependencies`.

That's a good point. Try something like this:
```js
function collectFoldersRecursive(packageJsonPath, foldersToCopy, symlinksToCopy) {
  const packageJsonFolderPath = path.dirname(getRealPath(packageJsonPath, symlinksToCopy));

  if (!foldersToCopy.has(packageJsonFolderPath)) {
    foldersToCopy.add(packageJsonFolderPath);
    // console.log(packageJsonFolderPath);

    const packageJson = JsonFile.load(packageJsonPath);

    // Union of keys from regular dependencies, peerDependencies, and optionalDependencies
    const allDependencyNames = new Set();
    // Just the keys from optionalDependencies
    const optionalDependencyNames = new Set();

    for (const name in packageJson.dependencies || {}) {
      allDependencyNames.add(name);
    }
    for (const name in packageJson.peerDependencies || {}) {
      allDependencyNames.add(name);
    }
    for (const name in packageJson.optionalDependencies || {}) {
      allDependencyNames.add(name);
      optionalDependencyNames.add(name);
    }

    for (const dependencyPackageName of allDependencyNames) {
      const resolvedDependency = resolve.sync(dependencyPackageName, {
        basedir: packageJsonFolderPath,
        preserveSymlinks: false, // since getRealPath() already implements this logic above
        packageFilter: (pkg, dir) => {
          // point "main" at a file that is guaranteed to exist
          // This helps resolve packages such as @types/node that have no entry point
          pkg.main = './package.json';
          return pkg;
        }
      });

      if (!resolvedDependency) {
        if (optionalDependencyNames.has(dependencyPackageName)) {
          // Ignore missing optional dependency
          continue;
        }
        throw new Error(`Error resolving ${packageName} from ${packageJsonPath}`);
      }

      const dependencyPackageJsonPath = packageJsonLookup.tryGetPackageJsonFilePathFor(resolvedDependency);
      if (!dependencyPackageJsonPath) {
        throw new Error(`Error finding package.json for ${resolvedDependency}`);
      }

      collectFoldersRecursive(dependencyPackageJsonPath, foldersToCopy, symlinksToCopy)
    }

  }
}
```

**Comment by poelstra** [comment]
Tried that method on my tree, and it's indeed as simple as that :)

(Note: I have slightly different code, so didn't test the exact code that you wrote, but we're using the same principle.)

**Comment by octogonz** [comment]
We really should build this into Rush.

**Comment by octogonz** [comment]
@iclanton and I were chatting about the subset deployment problem again today, since I recently encountered it at my job.  Our main insight was that it's pretty valuable to be able to:

- Carve out the subset of dependencies for a particular project in the monorepo
- Copy them to another computer where it will execute
- Have be an "XCopy deployment" that doesn't rely on any package manager or compiler tools (but maybe a simple script that fixes up symlinks)

This imposes constraints on your dependencies: If they are binaries, they must match your target machine OS and Node.js ABI contract.  If the package ran NPM `postinstall` scripts, they must not write files with absolute paths or other machine-specific assumptions. But imposing these dependencies greatly simplifies deployment.

This feature can be used to do lightweight deployments without Docker (my scenario).  And it could also be leveraged to make more efficient Docker images.

The script above already picks out the `FOLDERS TO COPY:` and `SYMLINKS TO COPY:`. I'm going to see if I can get it working for real...

**Comment by SoyYoRafa** [comment]
This is great! Imho, it shouldn't be rush responsibility to match target machine OS/node.js ABI if we want to orchestrate to a remote location. It should be the continuous integration tooling's responsibility to build a target that is consistent with production/remote environment.

**Comment by octogonz** [comment]
The way that `collectFoldersRecursive()` collects symlinks is not correct, because it ignores intermediary symlinks discovered during `resolve.sync()`.  That can be fixed using the `opts.realpathSync` hook to intercept folder traversal.  

The way that `getRealPath()` discovers links is also incorrect, because it does not consider paths that combine chains of symlinks.  Below is a prototype of a more complete symlink collector:

<details>
<summary>SymlinkAnalyzer.ts</summary>

```ts
import {
  FileSystem,
  FileSystemStats,
  Sort,
} from "@rushstack/node-core-library";
import * as path from "path";

interface IFileNode {
  kind: 'file';
  nodePath: string;
}

interface IFolderNode {
  kind: 'folder';
  nodePath: string;
}

interface ILinkNode {
  kind: 'link';
  nodePath: string;

  // If it is a symlink, the (immediate) target that the symlink resolves to.
  linkTarget: string;
}

type PathNode = IFileNode | IFolderNode | ILinkNode;

export interface ILinkInfo {
  sourcePath: string;
  targetPath: string;
  kind: 'fileLink' | 'folderLink';
}

export class SymlinkAnalyzer {
  // The directory tree discovered so far
  private readonly _nodesByPath: Map<string, PathNode> = new Map<string, PathNode>();

  // The symlinks that we encountered while building the directory tree
  private readonly _linkInfosByPath: Map<string, ILinkInfo> = new Map<string, ILinkInfo>();

  public analyzePath(inputPath: string, preserveLinks: boolean = false): PathNode {
    let pathSegments: string[] = path.resolve(inputPath).split(path.sep);
    let pathSegmentsIndex: number = 0;

    for (; ;) {
      const segment: string = pathSegments[pathSegmentsIndex];
      const currentPath: string = pathSegments.slice(0, pathSegmentsIndex + 1).join(path.sep);

      let currentNode: PathNode | undefined = this._nodesByPath.get(currentPath);
      if (currentNode === undefined) {
        const linkStats: FileSystemStats = FileSystem.getLinkStatistics(currentPath);

        if (linkStats.isSymbolicLink()) {
          const linkTargetPath: string = FileSystem.readLink(currentPath);
          const parentFolder: string = path.join(currentPath, '..');
          const resolvedLinkTargetPath: string = path.resolve(parentFolder, linkTargetPath);
          currentNode = {
            kind: 'link',
            nodePath: currentPath,
            linkTarget: resolvedLinkTargetPath
          };
        } else if (linkStats.isDirectory()) {
          currentNode = {
            kind: 'folder',
            nodePath: currentPath
          };
        } else if (linkStats.isFile()) {
          currentNode = {
            kind: 'file',
            nodePath: currentPath
          };
        } else {
          throw new Error('Unknown object type: ' + currentPath);
        }

        this._nodesByPath.set(currentPath, currentNode);
      }

      ++pathSegmentsIndex;

      if (!preserveLinks) {
        while (currentNode.kind === 'link') {
          const targetNode: PathNode = this.analyzePath(currentNode.linkTarget, true);

          // Have we created an ILinkInfo for this link yet?
          if (!this._linkInfosByPath.has(currentNode.nodePath)) {
            // Follow any symbolic links to determine whether the final target is a directory
            const targetIsDirectory: boolean = FileSystem.getStatistics(targetNode.nodePath).isDirectory();
            const linkInfo: ILinkInfo = {
              kind: targetIsDirectory ? 'folderLink' : 'fileLink',
              sourcePath: currentNode.nodePath,
              targetPath: targetNode.nodePath
            };
            this._linkInfosByPath.set(currentNode.nodePath, linkInfo);
          }

          const targetSegments: string[] = targetNode.nodePath.split(path.sep);
          const remainingSegments: string[] = pathSegments.slice(pathSegmentsIndex);
          pathSegments = [...targetSegments, ...remainingSegments];
          pathSegmentsIndex = targetSegments.length;
          currentNode = targetNode;
        }
      }

      if (pathSegmentsIndex >= pathSegments.length) {
        // We reached the end
        return currentNode;
      }

      if (currentNode.kind !== 'folder') {
        throw new Error('The path ends prematurely at: ' + inputPath);
      }
    }
  }

  public reportSymlinks(): ILinkInfo[] {
    const list: ILinkInfo[] = [...this._linkInfosByPath.values()];
    Sort.sortBy(list, x => x.sourcePath);
    return list;
  }
}

// Test it
const collector: SymlinkAnalyzer = new SymlinkAnalyzer();
collector.analyzePath("c:\\t\\a\\.\\b\\c\\d");
console.log(JSON.stringify(collector.reportSymlinks(), undefined, 2));
```

</details>

**Comment by jlsjonas** [comment]
@octogonz What seems to work for us to get all links properly resolved is to use tar with **both** the `--hard-dereference` and `--dereference` options (our use-case it to pipe it to a docker build).
The only issue with rush's current approach there is that there's no easy way to get rid of the devDependencies in the tree.
But when implementing into rush, I believe a lot of use-cases could be solved by simply having a _cleaned out_ tar.gz as a result.
Will try incorporating our strategy with your collection scripts and provide further feedback if feasible.

**Comment by octogonz** [comment]
@jlsjonas Right, the important problem here isn't replicating symlinks, but how to extract the subset of relevant files/folders to deploy.  Deploying a subset can:

- greatly reduce the footprint of files to be uploaded during deployment
- Remove problematic `devDependencies`. For example, a license/compliance issue may prevent them from being installed on a public server.

(Replicating symlinks came up as a subproblem for Approach 3 below.)

Suppose we have a monorepo like this, and we want to deploy `app1`:

- `app1`: depends on `ext-lib7` from NPM and `lib3`, with a dev dependency on `ext-tool8` and `tool6`
- `app2`: depends on `lib3` and `lib4`
- `lib3`: depends on `lib5`
- `lib4`: no dependencies
- `lib5`: peer dependency on `ext-lib7`
- `tool6`: no dependencies

In this example, the deployment subset should be `app1`, `ext-lib7`, `lib3`, and `lib5`.  We do NOT deploy `tool6` or `ext-tool8` because they are dev dependencies.

### Approach 1: Publishing to a private registry

The deployment subset is equivalent to what we'd get if we do this:

1. Build the monorepo
2. Publish all the local Rush projects to a temporary NPM registry such as [Verdaccio](https://www.npmjs.com/package/verdaccio)
3. Do  `pnpm install app1` in an empty `deploy/` folder, installing from the temporary registry. Because of our "XCopy deployment" goal, this step happens on the CI machine, not the web server.
4. Upload this `deploy/` folder to the deployment machine (preserving symlinks somehow) 

Besides the clumsiness of a temporary registry, the main downside is that the monorepo's `pnpm-lock.yaml` state is not used during installation.  I'm uncomfortable with that. We might not be deploying the same dependencies that we used during the build. If the monorepo has `pnpmfile.js` workarounds or other version tuning, that would have to be redone separately for the deployment lockfile. 

But otherwise Approach 1 is the most simpleminded and conventional solution.

### Approach 2: Using PNPM `--filter`

It turns out the latest PNPM workspace feature has pretty good support for partial installs. (Yeah PNPM! 🤘)  I found that this command will install the subset of `pnpm-lock.yaml` corresponding to our deployment subset:

```
pnpm install --filter app1... --prod
```

Note that it is an in-place installation; the local projects are symlinked and do not need to be published.

Anyway, we could imagine a `rush deploy` algorithm that proceeds like this:

1. For now, suppose we are using PNPM workspaces instead of `rush link` (Coming soon via https://github.com/microsoft/rushstack/issues/1887)
2. Clone the `master` branch, install, and build everything
3. Purge all `node_modules` folders that were installed in step 2
4. Run `pnpm install --filter app1... --prod` to install non-dev dependencies for `app1` only.
5. Calculate the "referenced projects" set to be `app1` plus all local projects that `app1` depends on
6. Build the XCopy file inventory by collecting the root `node_modules` folder, plus the "referenced projects" folders, and their untrimmed `node_modules` folders
7. To handle symlinks we could use `tar --hard-dereference --dereference` like you said. My approach was to make a `create-symlinks.js` postinstall script that would recreate them (to fix up relative links and NTFS junctions), but maybe `tar` can do that. (?)

Note that the symlink strategy here depends on an assumption:  ***PNPM's symlinks will never point outside the folders enumerated in step 6.***  That seems empirically true so far.

Since the local projects did not get "installed" by PNPM, there is no `.npmignore` filtering to keep `dist/blah.js` and discard `src/blah.ts`.  And we have to determine the "referenced projects" subset.  And we have to do something inefficient in steps 2+3.  

You could imagine a `pnpm deploy` feature that optimizes away these inefficiencies by more accurately simulating Approach 1 (without actually publishing local projects to a temporary NPM registry).  Delegating this job to PNPM is seems best long term, since this fundamentally feels like a package manager problem.  

Even if PNPM doesn't give us this feature, there may be clever ways to trick it into being more optimal, e.g. by invoking its API, or hooking into **pnpmfile.js,** or rewriting **package.json** files to use `file:` version specifiers instead of Verdaccio.

### Approach 3: The `collectFolders()` script

This is the approach I was prototyping over the weekend. The algorithm works like this:

1. Start in the `app1` folder and call `resolve.sync()` for each of its dependencies, then recursively do this for all the indirect dependencies
2. Along the way, collect the list of package.json folders we visited
3. Along the way, collect an inventory of symlinks that we traversed during module resolution
4. Build the XCopy file inventory as the set of folders from step 2, minus their node_modules subfolders (except for explicitly visited subfolders)

This is very fast, and it is compatible with both `rush link` and PNPM workspaces. 

The big downside is that we're trusting `package.json` to tell us which dependencies are actually needed, when everyone knows there are lots and lots of [phantom dependencies](https://rushjs.io/pages/advanced/phantom_deps/) in the registry.  `pnpm install` employs various hoisting heuristics to mitigate that.  My bet was that these phantoms primarily plague complex compile-time software like Jest and Webpack and Babel, which generally do NOT get deployed.  So perhaps the dependencies of a web service are relatively well-behaved.  And when they are not, perhaps there is a somewhat easy way to fix it up by including more folders (based on PNPM's `node_modules` tree).  (?)  I'm still experimenting with that. I'll follow up with the results.

**Comment by octogonz** [comment]
😃 I finished a prototype of Approach 3. So far it works great.  I was able to successfully deploy Rush and a Node.js web server (using `express`) without any troubles. Everything just worked on the first try, no phantom dependency issues.

Tomorrow I will try converting this prototype into a PR that we can test more thoroughly.

**Comment by octogonz** [comment]
## Proposed design
The deployment is controlled by JSON files under the **common/config/deploy-scenarios** folder, one file for each scenario.  The file names must be `lower-case-with-dashes.json`.

Example file structure:

**common/config/deploy-scenarios/aws.json**
```ts
{
  /**
   * When deploying a local Rush project, by the "devDependencies" are normally excluded.
   * Set this to "true" to include them.
   *
   * The default value is false.
   */
   "includeDevDependencies": false,

  /**
   * When deploying a local Rush project, by the .npmignore filter is normally applied, so the included files are
   * what would be packaged by "npm pack".  Set this to "true" to include all files (except for the "node_modules"
   * folder).
   *
   * The default value is false.
   */
   "includeNpmIgnoreFiles": false,

  /**
   * The list of local Rush projects to be included in the deployment.  For each project, the "rush deploy" operation
   * will recursively include the package.json dependencies which may be NPM packages or other local Rush projects.
   */
  "includedProjects": [
    {
      /**
       * The package name for the project, which must be defined in rush.json.
       */
      "projectName": "web-server"
    }
  ]
}
```

The file can be scaffolded using a new command:
```bash
# Generates an empty template "common/config/deploy-scenarios/aws.json"
$ rush init-deploy --scenario aws
```

The deployment is invoked like this:
```bash
# By default, the output is written to the target folder "common/deploy" regardless
# of the scenario.  If the folder is not empty, an error occurs, unless "--overwrite" is
# specified, in which case the previous contents are deleted.
$ rush deploy --scenario aws --overwrite

# Or you can specify a custom target folder:
$ rush deploy --scenario aws --overwrite --target-folder D:\MyDeployment
```

(We considered deploying to **common/temp/deploy** by default, but there is a phantom dependency concern due to **common/temp/node_modules**.)

**Comment by octogonz** [comment]
Feedback is welcome regarding this design. :-)

**Comment by SoyYoRafa** [comment]
Can there be a command to create an independent package for all packages that are applications in the monorepo? We would want to create a package for each application in the monorepo. I think having to create a new file for each new app is bit more verbose/slow. It would be the responsibility of an orchestration tool to figure out what needs to be deployed in which environment.

**Comment by octogonz** [comment]
@SoyYoRafa I'm not sure what you mean. Could you give an example?

Above I started an example like this:

> Suppose we have a monorepo like this, and we want to deploy `app1`:
>
> - `app1`: depends on `ext-lib7` from NPM and `lib3`, with a dev dependency on `ext-tool8` and `tool6`
> - `app2`: depends on `lib3` and `lib4`
> - `lib3`: depends on `lib5`
> - `lib4`: no dependencies
> - `lib5`: peer dependency on `ext-lib7`
> - `tool6`: no dependencies

Are you asking to specify something like:

```js
"projectsToDeploy":  [ "app1", "app2" ]
```
...and then `rush deploy` would produce:

-  `common/deploy/app1` with only `app1`, `ext-lib7`, `lib3`, `lib5`
-  `common/deploy/app2` with only `app2`, `lib3`, `lib4`, `lib5`

i.e. two entire copied subsets?

An example would be really helpful.

**Comment by octogonz** [comment]
🎈 Here's a draft PR with my current progress:  https://github.com/microsoft/rushstack/pull/1898 

This integrates my prototype into a proper "rush deploy" command as proposed above.  To try it out, the PR description has usage instructions.  

I would really appreciate feedback on whether this successfully deploys real world projects or not.

**Comment by wbern** [comment]
> Feedback is welcome regarding this design. :-)

This might not be the feedback you're expecting but, someday in the future, our org is going to need to re-download the files we deployed for local use, to avoid lengthy rush update and rush build step. Excluding dependencies that aren't necessary will help download completion time of such a feature, and we would not have to rely on the npm registry to get that small bundle.

My take above is likely not a 100% fit with the proposed PR, for example I don't know how the symlinking could be done after downloading a deployed set of packages to the packages already in the repo, so maybe that could be worth your consideration while you're working on this.

**Comment by buffcode** [comment]
> Can there be a command to create an independent package for all packages that are applications in the monorepo? We would want to create a package for each application in the monorepo.

We have almost the same requirement:
- we have ~30 subprojects in our monorepo
- our CI builds the main projects (~5) in separate jobs
- those main projects will be deployed

In order to build them with the current approch, one would have to create 5 deployment configurations which only differ in `includedProjects: [{"projectName": "projectX"}]`.

Of course this could be auto-generated or so.

Any way, as the first MVC I think this is a suitable solution and can be enhanced in further iterations.

**Comment by octogonz** [comment]
> This might not be the feedback you're expecting but, someday in the future, our org is going to need to re-download the files we deployed for local use, to avoid lengthy rush update and rush build step. Excluding dependencies that aren't necessary will help download completion time of such a feature, and we would not have to rely on the npm registry to get that small bundle. (main drawbacks of npm registry for us is complicated use and the requirement that all packages need to be published in order to be downloadable, which is not always an option in non-master branches)
> 
> My take above is likely not a 100% fit with the proposed PR, for example I don't know how the symlinking could be done after downloading a deployed set of packages to the packages already in the repo, so maybe that could be worth your consideration while you're working on this.

@wbern for my own scenario, `rush deploy` will not create the symlinks immediately, but instead emit a `create-symlinks.js` script that you run on the remote server.  I'm thinking the data will be in a companion file, maybe named `deploy-symlinks.json`.  

Perhaps you could use this feature to solve your problem:
- Save a copy of the folder prior to running the script, and that can be the thing you download; OR
- Use a download method that skips symlinks, and then re-run the script on your local machine

**Comment by octogonz** [comment]
Here is a revised config file design that introduces a "subdivisions" concept for multi-deployments.

**common/config/deploy-scenarios/aws.json**
```ts
{
  /**
   * When deploying a local Rush project, by the "devDependencies" are normally excluded.
   * Set this to "true" to include them.
   *
   * The default value is false.
   */
  "includeDevDependencies": false,

  /**
   * When deploying a local Rush project, by the .npmignore filter is normally applied, so the included files are
   * what would be packaged by "npm pack".  Set this to "true" to include all files (except for the "node_modules"
   * folder).
   *
   * The default value is false.
   */
  "includeNpmIgnoreFiles": false,

  /**
   * Specify how symlinks will be created in the deployed folder:
   *
   * - "default": Create the symlinks after copying the files; this is the default behavior.
   * - "script": A Node.js script called "create-symlinks.js" will be written.  When executed, this script will
   *   create the symlinks described in the "deploy-metadata.json" output file.
   * - "none": Do nothing; some other tool may create them later.
   */
  "symlinkCreation": "script",

  /**
   * Customize how local Rush projects are processed, if they are included in the deployment.
   */
  "projectSettings": [
    {
      /**
       * The full package name of the project, which must be declared in rush.json.
       */
      "projectName": "@microsoft/api-extractor",

      /**
       * A list of additional local Rush projects to be deployed with this project (beyond the package.json
       * dependencies).  Specify full package names, which must be declared in rush.json.
       */
      "additionalProjectsToDeploy": [
        "@microsoft/api-documenter"
      ]
    }
  ],

  /**
   * This optional setting enables multiple deployments to be prepared as a single scenario.  The "subdivisions"
   * setting is an array of Rush project names.  These names can omit the NPM scope if no ambiguity would arise.
   * The subdivision name will be used to create a subfolder under the deployment target folder.  Each
   * subdivison folder will receive its own copy of the associated project and its dependencies.
   *
   * If the "subdivisions" setting is omitted, then all projects from the "projectSettings" will be
   * deployed as a single set, without creating a subdivision folder.
   */
  "subdivisions": [ "api-extractor", "repo-toolbox" ]
}
```
In this example, `rush deploy` would output:
```bash
# subdivision #1
common/deploy/api-extractor/apps/api-extractor/...          # via subdivisions setting
common/deploy/api-extractor/apps/api-documenter/...         # via additionalProjectsToDeploy
common/deploy/api-extractor/libraries/node-core-library/... # local dependency
common/deploy/api-extractor/common/temp/node_modules/...

# subdivision #2
common/deploy/repo-toolbox/repo-scripts/repo-toolbox/...   # via subdivisions setting
common/deploy/repo-toolbox/libraries/node-core-library/... # local dependency
common/deploy/repo-toolbox/common/temp/node_modules/...
```
Whereas if we comment out the "subdivisions" setting, then `rush deploy` will not create the subdivision folders:
```bash
common/deploy/apps/api-extractor/...          # via projectSettings
common/deploy/apps/api-documenter/...         # via additionalProjectsToDeploy
common/deploy/libraries/node-core-library/... # local dependency
common/deploy/common/temp/node_modules/...
```

(@iclanton has also suggested that `rush deploy` could flatten the monorepo folder structure, i.e. eliminating the `apps` and `libraries` levels. This is not too difficult to implement, but outside the scope of the MVP.)

**Comment by octogonz** [comment]
"subdeployments" is maybe a better name than "subdivisions"

**Comment by octogonz** [comment]
I have updated the PR to implement this new design.

**Comment by iclanton** [comment]
I tested this last night with a tool project that depends on other projects inside a large monorepo and it works as expected. Here's what I did:

- created a deploy configuration
- ran `rush deploy --scenario repo-toolbox` (my project actually is called `repo-toolbox`)
- invoked the `start.js` script from the tool inside the `common/deploy/repo-toolbox` folder.

This correctly printed the help message for the tool.

I think it's a little awkward that the whole monorepo structure is maintained, especially if you're deploying only a single project. It'd also be really useful if what ended up in the `deploy` folder was a tarball (or perhaps a tarball in addition to a folder) that you can copy to another machine, expand, and invoke without needing to do anything else (or with only having to do a minimal amount of work involving no network calls).

**Comment by octogonz** [comment]
I simplified the config file a bit and promoted the draft PR.  I marked the feature as "experimental", so we can merge+publish it so people can test it out more easily.

**Comment by octogonz** [comment]
🎈 This feature was published as Rush `5.24.0`. I'm really excited to finally get this solved! Thanks @arikon, @iclanton, and everyone else who helped out!  

If anyone has further feedback/ideas, please let us know in a new GitHub issue.
